lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -0,0 +1,909 @@
+  llvm::SmallVector<ModulePort, 8> ports;
+
+  // Add all inputs of funcOp.
+  int args_idx = 0;
+  for (auto &arg : funcOp.getArguments()) {
+    StringAttr portName =
+        rewriter.getStringAttr("arg" + std::to_string(args_idx));
+    FIRRTLType bundlePortType = getBundleType(arg.getType(), /*isFlip=*/false);
+    ports.push_back(ModulePort(portName, bundlePortType));
+    args_idx += 1;
+  }
+
+  // Add all outputs of funcOp.
+  for (auto portType : funcOp.getType().getResults()) {
+    StringAttr portName =
+        rewriter.getStringAttr("arg" + std::to_string(args_idx));
+    FIRRTLType bundlePortType = getBundleType(portType, /*isFlip=*/true);
+    ports.push_back(ModulePort(portName, bundlePortType));
+    args_idx += 1;
+  }
+
+  // Create a FIRRTL module, and inline the funcOp into it.
+  FModuleOp topModuleOp = rewriter.create<FModuleOp>(
+      funcOp.getLoc(), rewriter.getStringAttr(funcOp.getName()),
+      ArrayRef<ModulePort>(ports));
+  rewriter.inlineRegionBefore(funcOp.getBody(), topModuleOp.getBody(),
+                              topModuleOp.end());
+
+  // Merge the second block (inlined from funcOp) of the top-module into the
+  // entry block.
+  auto blockIterator = topModuleOp.getBody().begin();
+  Block *entryBlock = &*blockIterator;
+  Block *secondBlock = &*(++blockIterator);
+
+  // Replace uses of each argument of the second block with the corresponding
+  // argument of the entry block.
+  for (int i = 0, e = secondBlock->getNumArguments(); i < e; ++i) {
+    BlockArgument oldArgument = secondBlock->getArgument(i);
+    BlockArgument newArgument = entryBlock->getArgument(i);
+    oldArgument.replaceAllUsesWith(newArgument);
+  }
+
+  // Move all operations of the second block to the entry block.
+  while (!secondBlock->empty()) {
+    Operation &op = secondBlock->front();
+    op.moveBefore(entryBlock->getTerminator());
+  }
+  rewriter.eraseBlock(secondBlock);
+
+  return topModuleOp;
+}
+
+//===----------------------------------------------------------------------===//
+// FIRRTL Sub-module Related Functions
+//===----------------------------------------------------------------------===//
+
+/// Check whether a submodule with the same name has been created elsewhere.
+/// Return the matched submodule if true, otherwise return nullptr.
+FModuleOp checkSubModuleOp(FModuleOp topModuleOp, Operation &oldOp) {
+  for (auto &op : topModuleOp.getParentRegion()->front()) {
+    if (auto subModuleOp = dyn_cast<FModuleOp>(op)) {
+      if (StringRef(getSubModuleName(oldOp)) == subModuleOp.getName()) {
+        return subModuleOp;
+      }
+    }
+  }
+  return FModuleOp(nullptr);
+}
+
+/// All standard expressions and handshake elastic components will be converted
+/// to a FIRRTL sub-module and be instantiated in the top-module.
+FModuleOp createSubModuleOp(FModuleOp topModuleOp, Operation &oldOp,
+                            ConversionPatternRewriter &rewriter) {
+  rewriter.setInsertionPoint(topModuleOp);
```

Comment:
Perhaps this should be a custom builder for SubModuleOp that takes lists of types?

Reply:
Yes, this is possible. The current builder of FModuleOp is taking a list of {portName, portType} pairs. Not sure whether this createSubModuleOp() method is common enough for FIRRTL dialect users to become a custom builder of FModuleOp.

Reply:
This method is especially useful when lowering high-level IRs to FIRRTL IR.

Reply:
+1, it would be nice to add this as a method on FModuleOp.

Reply:
👌 Will do that.